src/Core/OrganizationFeatures/OrganizationSubscriptionUpdate/UpdateSecretsManagerSubscriptionCommand.cs

```diff
@@ -0,0 +1,354 @@
+        }
+
+    }
+
+    private async Task<string> AdjustSeatsAsync(Organization organization, int seatAdjustment)
+    {
+        if (organization.SmSeats == null)
+        {
+            throw new BadRequestException("Organization has no Secrets Manager seat limit, no need to adjust seats");
+        }
+
+        if (string.IsNullOrWhiteSpace(organization.GatewayCustomerId))
+        {
+            throw new BadRequestException("No payment method found.");
+        }
+
+        if (string.IsNullOrWhiteSpace(organization.GatewaySubscriptionId))
+        {
+            throw new BadRequestException("No subscription found.");
+        }
+
+        var plan = StaticStore.SecretManagerPlans.FirstOrDefault(p => p.Type == organization.PlanType);
+
+        if (plan == null)
+        {
+            throw new BadRequestException("Existing plan not found.");
+        }
+
+        if (!plan.HasAdditionalSeatsOption)
+        {
+            throw new BadRequestException("Plan does not allow additional Secrets Manager seats.");
+        }
+
+        var newSeatTotal = organization.SmSeats.HasValue
+            ? organization.SmSeats.Value + seatAdjustment
+            : seatAdjustment;
+
+        if (plan.BaseSeats > newSeatTotal)
+        {
+            throw new BadRequestException($"Plan has a minimum of {plan.BaseSeats} Secrets Manager  seats.");
+        }
+
+        if (newSeatTotal <= 0)
+        {
+            throw new BadRequestException("You must have at least 1 Secrets Manager seat.");
+        }
+
+        var additionalSeats = newSeatTotal - plan.BaseSeats;
+
+        if (plan.MaxAdditionalSeats.HasValue && additionalSeats > plan.MaxAdditionalSeats.Value)
+        {
+            throw new BadRequestException($"Organization plan allows a maximum of " +
+                                          $"{plan.MaxAdditionalSeats.Value} additional Secrets Manager seats.");
+        }
+
+        if (!organization.SmSeats.HasValue || organization.SmSeats.Value > newSeatTotal)
+        {
+            var occupiedSeats = await _organizationUserRepository.GetOccupiedSmSeatCountByOrganizationIdAsync(organization.Id);
+            if (occupiedSeats > newSeatTotal)
+            {
+                throw new BadRequestException($"Your organization currently has {occupiedSeats} seats filled. " +
+                                              $"Your new plan only has ({newSeatTotal}) seats. Remove some users.");
+            }
+        }
+
+        var paymentIntentClientSecret = await _paymentService.AdjustSeatsAsync(organization, plan, additionalSeats);
+
+        await _referenceEventService.RaiseEventAsync(
+            new ReferenceEvent(ReferenceEventType.AdjustSmSeats, organization, _currentContext)
+            {
+                Id = organization.Id,
+                PlanName = plan.Name,
+                PlanType = plan.Type,
+                Seats = newSeatTotal,
+                PreviousSeats = organization.SmSeats
+            });
```

Comment:
This needs to be done last, e.g. when you call `ReplaceAndUpdateCache`. The idea is that we want to validate everything before we write to db or update Stripe.

Maybe you can move these calls into private methods like `ProcessCharges` and `RaiseEvents`, to keep the main public method nice and simple.

Reply:
@eliykat , I am not very clear with this comment

Reply:
done the implementation. Please check